sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,162 @@
+    private final SynchronousReceiveWork initialWork;
     private volatile Subscription subscription;
 
-    SynchronousMessageSubscriber(SynchronousReceiveWork work) {
-        this.work = Objects.requireNonNull(work, "'work' cannot be null.");
+
+    private SynchronousReceiveWork currentWork;
+    private Disposable timeoutOperation;
+    private Disposable drainQueueDisposable;
+
+    SynchronousMessageSubscriber(long prefetch, SynchronousReceiveWork initWork) {
+        this.prefetch = prefetch;
+        this.initialWork = initWork;
     }
 
     /**
      * On an initial subscription, will take the first work item, and request that amount of work for it.
-     *
      * @param subscription Subscription for upstream.
      */
     @Override
     protected void hookOnSubscribe(Subscription subscription) {
-        if (this.subscription == null) {
-            this.subscription = subscription;
-        }
-
-        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
-            work.getTimeout());
+        this.subscription = subscription;
 
-        subscription.request(work.getNumberOfEvents());
+        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", initialWork.getId(),
+            initialWork.getNumberOfEvents(), initialWork.getTimeout());
 
-        timer.schedule(new ReceiveTimeoutTask(work.getId(), this::dispose), work.getTimeout().toMillis());
+        // This will trigger subscription.request(N) and queue up the work
+        queueWork(initialWork);
     }
 
     /**
      * Publishes the event to the current {@link SynchronousReceiveWork}. If that work item is complete, will dispose of
      * the subscriber.
-     *
-     * @param value Event to publish.
+     * @param message Event to publish.
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
+        currentWork.next(message);
 
-        if (work.isTerminal()) {
-            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", work.getId());
-            dispose();
+        if (currentWork.isTerminal()) {
+            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", currentWork.getId());
+            completeCurrentWork(currentWork);
         }
     }
 
-    @Override
-    protected void hookOnComplete() {
-        logger.info("[{}] Completed. No events to listen to.", work.getId());
-        dispose();
+    private void completeCurrentWork(SynchronousReceiveWork currentWork) {
+
+        if (isTerminated()) {
+            return;
+        }
+
+        currentWork.complete();
+        logger.verbose("[{}] work completed.", currentWork.getId());
+
+        if (timeoutOperation != null && !timeoutOperation.isDisposed()) {
```

Comment:
This can be inside the hookOnNext method. And return if the currentWork is not terminal to save the nested if.